src/collections/Form/Form.js

```diff
@@ -1,143 +1,264 @@
+    const name = getNodeName(node)
+    const checkboxesByName = _.filter(arr, { name })
 
-  serializeJson = () => {
-    const form = this.refs.element
-    const json = {}
-
-    _.each(['input', 'textarea', 'select'], (tag) => {
-      _.each(form.getElementsByTagName(tag), (node, index, arr) => {
-        const name = node.getAttribute('name')
-
-        switch (node.getAttribute('type')) {
-          case 'checkbox':
-            json[name] = { checked: node.checked }
-            break
-
-          case 'radio':
-            json[name] = {
-              value: _.find(arr, { name, checked: true }).value,
-            }
-            break
-
-          default:
-            json[name] = { value: node.value }
-            break
-        }
-      })
-    })
-
-    return json
-  }
+    // single: (value|checked)
+    if (checkboxesByName.length === 1) {
+      json[name] = node.checked && node.value !== 'on' ? node.value : node.checked
+      debugSerializedResult(json, name, node)
+      return
+    }
+
+    // groups (checked): [value, ...]
+    if (!Array.isArray(json[name])) json[name] = []
+    if (node.checked) json[name].push(node.value)
+    debugSerializedResult(json, name, node)
+
+    // in dev, warn about multiple checkboxes with a default browser value of "on"
+    if (process.NODE_ENV !== 'production' && node.value === 'on') {
+      const errorMessage = [
+        "Encountered a checkbox in a group with the default browser value 'on'.",
+        'Each checkbox in a group should have a unique value.',
+        "Otherwise, the checkbox value will serialize as ['on', ...].",
+      ].join(' ')
+      console.error(errorMessage, node, formNode) // eslint-disable-line no-console
+    }
+  })
+
+  // ----------------------------------------
+  // Radios
+  // checked: { name: checked value }
+  // none:    { name: null }
+
+  _.each(formNode.querySelectorAll('input[type="radio"]'), (node, index, arr) => {
+    const name = getNodeName(node)
+    const checkedRadio = _.find(arr, { name, checked: true })
+
+    if (checkedRadio) {
+      json[name] = checkedRadio.value
+    } else {
+      json[name] = null
+    }
+
+    debugSerializedResult(json, name, node)
+
+    // in dev, warn about radios with a default browser value of "on"
+    if (process.NODE_ENV !== 'production' && node.value === 'on') {
+      const errorMessage = [
+        "Encountered a radio with the default browser value 'on'.",
+        'Each radio should have a unique value.',
+        "Otherwise, the radio value will serialize as { [name]: 'on' }.",
+      ].join(' ')
+      console.error(errorMessage, node, formNode) // eslint-disable-line no-console
+    }
+  })
+
+  // ----------------------------------------
+  // Other inputs
+  // { name: value }
+
+  _.each(formNode.querySelectorAll('input:not([type="radio"]):not([type="checkbox"])'), (node) => {
+    const name = getNodeName(node)
+    json[name] = node.value
+    debugSerializedResult(json, name, node)
```

Comment:
It's a noop only called onSubmit.  Though, we could wrap each call in a production check.  I don't think this will ever show up in a perf profile, but I can wrap them all if you feel very strongly about it still.